### DISCUSSION TITLE
Blocking Rules in EM for dedupe_only

### DISCUSSION THREAD

**Discussion by mshearer0** [moj-analytical-services/splink#1518] [General]
I have a dataset of ~10M records with the following structure:

      Column          Dtype 
 0   Lastname        string
 1   Middlename      string
 2   Firstname       string
 4   Year            int64 
 5   Month           int64 
 6   Title           string
 7   Nationality     string
 8   unique_id       int64 

I'm trying to dedupe_only (using jarow on Firstname, Middlename and exact matches otherwise) so i can cluster and illustrate the variations in attributes of different entities that get clustered together. Prediction (blocking on exact matching Lastname, Month and Year) works well as does clustering. However at the full scale I haven't been able to find a combination of blocking rules that will allow EM to execute in a reasonable timeframe. I have used higher memory and disk options on GCP but not yet tried with Spark.

As a workaround i've used:

linker.estimate_parameters_using_expectation_maximisation("l.Lastname = r.Lastname and l.Month = r.Month and l.Title = r.Title and l.Nationality = r.Nationality and **l.Month =6**", fix_u_probabilities=False)
linker.estimate_parameters_using_expectation_maximisation("l.Firstname = r.Firstname and l.Year = r.Year and l.Middlename = r.Middlename and **l.Year = 1970**", fix_u_probabilities=False)

Specifying a month and year respectively cuts down the number of samples dramatically but clearly a truely random sample to reduce volume would be more representative and produce better m and u values (as per discussion https://github.com/moj-analytical-services/splink/discussions/1075)

Is there a better way of EM blocking to reduce volume and allow the EM process to execute?

**Comment by RobinL** [comment]
So just to check - this is DuckDB?

Do you know why it's taking a long time?  Is each iteration taking too long, or is it failing to converge?

Have you tried counting how many comparisons will be created by your proposed blocking rule?  In DuckDB, you'll probably want to try to keep it below about 10m if possible.  You can use [this function](https://moj-analytical-services.github.io/splink/linkerbloc.html?h=count_n#splink.linker.Linker.count_num_comparisons_from_blocking_rule).

To see whether it's a converge problem, check the logs and/or set https://moj-analytical-services.github.io/splink/settings_dict_guide.html#max_iterations lower

**Reply by mshearer0** [reply]
Hi Robin. Yes, DuckDB. It isn’t getting to the iterations, seems to be writing huge volumes of data (100GB) to disk before first iteration. I’ve haven’t tried to measure the number of comparisons generated by the blocking rules (I will do now) but I expect it’s well over 10M. 

Blocking on Lastname, Month, Year will execute EM successfully after a few minutes but then I’m left with less strong blockers to use to generate m and u values for Lastname, Month and Firstname. This step doesn’t complete. Initially it failed with kernel errors which I solved by adding more memory. Then it ran out of disk space so I added more of that too! 



.

**Reply by mshearer0** [reply]
linker.count_num_comparisons_from_blocking_rule("l.Lastname = r.Lastname and l.Month = r.Month and l.Title = r.Title and l.Nationality = r.Nationality") gives 777M comparisions.

linker.count_num_comparisons_from_blocking_rule("l.Firstname = r.Firstname and l.Year = r.Year and l.Middlename = r.Middlename")
gives 69M.

The only option i can see is to run the EM process on a subset of the data and then use those values to predict on the full set?
